File: c++/N-th Tribonacci Number.cpp
#include <iostream>
#include <vector>

class Solution {
public:
    int tribonacci(int n) {
        if (n == 0) return 0;
        if (n == 1 || n == 2) return 1;
        int t0 = 0, t1 = 1, t2 = 1;
        for (int i = 3; i <= n; ++i) {
            int tn = t0 + t1 + t2;
            t0 = t1;
            t1 = t2;
            t2 = tn;
        }
        return t2;
    }
};

// test case
int main() {
    Solution sol;
    int n1 = 4;
    int result1 = sol.tribonacci(n1);
    std::cout<<"result :"<<result1<<std::endl;
    return 0;
}

File: python/N-th Tribonacci Number.py
class Solution:
    def tribonacci(self, n: int) -> int:
        if n == 0:
            return 0
        if n == 1 or n == 2:
            return 1
        t0, t1, t2 = 0, 1, 1
        for i in range(3, n + 1):
            tn = t0 + t1 + t2
            t0, t1, t2 = t1, t2, tn
        return t2

# Test cases
if __name__ == "__main__":
    solution = Solution()
    print(solution.tribonacci(4))  # Output: 4
    print(solution.tribonacci(25)) # Output: 1389537